scripts/sync_template_tracker.py

```diff
@@ -1,122 +1,212 @@
-        thread_q.append(thread)
-        thread.start()
-
-    # Join the queued calls
-    for thread in thread_q:
-        thread.join()
-
-    seen_templates = set()
-
-    if mark_usable is None:
-        usable = {}
-    else:
-        usable = {'usable': mark_usable}
-
-    existing_provider_templates = [
-        pt['id']
-        for pt
-        in trackerbot.depaginate(api, api.providertemplate.get())['objects']]
-
-    # Find some templates and update the API
-    for template_name, providers in template_providers.items():
-        template_name = str(template_name)
+def parse_cmdline():
+    parser = trackerbot.cmdline_parser()
+    parser.add_argument(
+        '--mark-usable',
+        default=None,
+        action='store_true',
+        help="Mark all added templates as usable",
+    )
+    parser.add_argument(
+        '--provider-key',
+        default=None,
+        dest='selected_provider',
+        nargs='*',
+        help='A specific provider key to sync for',
+    )
+    parser.add_argument(
+        '--outfile',
+        dest='outfile',
+        default=log_path.join('sync_template_tracker_report.log').strpath,
+        help='Output file for tabulated reports on ProviderTemplate actions'
+    )
+    parser.add_argument(
+        '--verbose',
+        default=False,
+        action='store_true',
+        help='Log to stdout'
+    )
+    args = parser.parse_args()
+    return dict(args._get_kwargs())
+
+
+def main(trackerbot_url, mark_usable=None, selected_provider=None, **kwargs):
+    tb_api = trackerbot.api(trackerbot_url)
+
+    all_providers = set(
+        selected_provider
+        or [key for key, data in cfme_data.management_systems.items()
+            if 'disabled' not in data.get('tags', [])])
+
+    bad_providers = manager.Queue()
+    # starmap the list of provider_keys into templates_on_provider
+    # return is list of ProvTemplate tuples
+    with ThreadPool(8) as pool:
+        mgmt_templates = pool.starmap(
+            templates_on_provider,
+            ((provider_key, bad_providers) for provider_key in all_providers)
+        )
+
+    # filter out the misbehaving providers
+    bad_provider_keys = []
+    while not bad_providers.empty():
+        bad_provider_keys.append(bad_providers.get())
+    logger.warning('Filtering out providers that failed template query: %s', bad_provider_keys)
+    working_providers = set([key for key in all_providers if key not in bad_provider_keys])
+
+    # Flip mgmt_templates into dict keyed on template name, listing providers
+    # [
+    #   {prov1: [t1, t2]},
+    #   {prov2: [t1, t3]},
+    # ]
+    #
+    # template_providers should look like:
+    # {
+    #   t1: [prov1, prov2],
+    #   t2: [prov1],
+    #   t3: [prov2]
+    # }
+    mgmt_providertemplates = defaultdict(list)
+    for prov_templates in mgmt_templates:
+        # expecting one key (provider), one value (list of templates)
+        if prov_templates is not None:
```

Comment:
for key, value in prov_templates.items() ?

Reply:
Need a little bit more than just this, to handle empty results, but yeah I like this better than using next(iter()).